### ISSUE TITLE
Why do things work like this?

### ISSUE THREAD

**Issue by mattapperson** [eridem/patata#6]
As I am working on pushing this forward, I am just curious as to a few engineering choices.

1) Why are all items required and then passed around as the deps variable? this makes it really hard to follow the code imho vs the normal process of just require what ya need.

2) Why is there a patata.yaml, and a config/config.yaml... both are required or are created yes none of the values are truly needed? 

3) I get a default, but can we make the path to features a config option?

4) The log.exit method makes tracking down issues really hard due to lack of stack trace... can we improve on that?

5) The breakdown of features to step definitions is great on one hand, but I know for me and my team we would rather have JUST global, generic definitions, and only use specific ones if needed... Can we support that mindset here too?

6) init command is really cool, but if there is already a package.json can't we have it just add whats needed to the existing project?

Thank yo very much for indulging me here. Once I understand these items better I will be happy to help implement what needs done and PR it

**Comment by eridem** [comment]
Hi!

Here some answers.

1. I use `deps` as an easy option of dependency injection. All initialization are done on the `index.js` as bootstrap file. Doing that I avoid the overuse of `require` in each file. This may be my way to develop, maybe we can have further discussions on this if you guys are more involved in the future :-) Do not have any problem to improve the code :-)
2. `patata.yml` is the configuration for the engine. For instance, to set up HockeyApp token (<https://patataio.github.io/#integrationshockeyapp>) or other Patata settings (the CLI tool). `config/config.yml` are configurations for the test cases (<https://patataio.github.io/#filestructureconfig>)
3. Patata uses Cucumber and Appium. My idea with the project was to avoid have discussions about how to organize the file system and the structure of the project. Making an standard and something quick and easy to set up. The first versions on the projects were a bit more flexible with a lot of settings, but after some releases I figured out that the flexibility brought more troubles to the team for understanding and with conflicts. I decided to remove unnecessary settings and create a solid static structure based on feedback. Then I improved more the CLI tool with error messages and help. Anyway, I am open for changes, but I think we should keep an static way to do things and improve scaffolding/messages/errors.
4. Sure. Any improvement is welcome.
5. You can decide to have global definitions if you wish now. Patata reads all files previous to their run. In our company, we made global some GIVEN steps related with logins. Even it is possible, I promote to isolate features as much as possible, but sometimes is needed. Do you have any suggestion about how we could support both mindset without messing up? All ideas are welcome :-)
6. Sure thing! Excellent idea.

I am happy to hear you wanna be involved! And it could be great to receive any PR from you, there are a lot of improvements I would like to add too that we could discuss if there are beneficial for your team.

**Comment by mattapperson** [comment]
Ok so some follow ups (thinking I should have had multiple tickets now to make this easier lol)

1. OK, I get where your coming from, I think there is room for improvement but for now just getting this somewhere where it works for my team is more important to me then cleaning that up so we can leave it be.

2. I am embarrassed now because a saw this earlier and forgot... my bad

3. I think my concern here is so many extra directories off of root make it hard to read... can we just point it to `./integration-tests/*`?

5. I was thinking of just adding an extra default search paths of:
```js
      .concat(glob.sync(`features/definitions.${platform}.js`))
      .concat(glob.sync(`features/definitions.common.js`))
```
Though ideally, imho definitions belong in their own folder... at the same level of features so more ideal to be would be:
```js
      .concat(glob.sync(`definitions/default.common.js`))
      .concat(glob.sync(`definitions/default.${platform}.js`))
      .concat(glob.sync(`definitions/${platform}.js`))
      .concat(glob.sync(`definitions/common.${platform}.js`))
```

As for Pos, you bet ya! We are betting the future of our company on improving quality, and integration tests we feel are key to that... so we are very vested in this idea

**Comment by mattapperson** [comment]
Actually, for 2... but why are they required? in particular the `patata.yml`... sure use it... but we should be able to run without it right?

**Comment by eridem** [comment]
Hi @mattapperson,

For **2** and **3**, have you considered the possibility to use the command `patata --cwd [PATH] [COMMANDS/OPTIONS]`. So you can set the tests any place you wish?

You could update your `package.json` project with a script similar to this:

```
"scripts": {
  "patata": "patata --cwd ./integration-tests"
}
```

And run it like

```
npm run patata [COMMAND/OPTIONS]
```

(I did not test it using it on the `scripts` section of `package.json`, but I believe it should work)

For **2**, the only reason to have `patata.yml` as mandatory file is for the CLI tool to realize that you are inside a `patata` project. Otherwise, commands like `patata run ...` will give you an error telling you that you are not inside a patata `project`. This can be solved with `patata --cwd [PATH] [COMMANDS/OPTIONS]` too. If we are able to make the CLI understand where the `patata` project is, we may not need this file in the future.

If you think my comments does not work for you, let's keep the discussion open :-)

For **4**, let me see if I understand you correctly. Here an example of the file system I understood from you:

```
+ root
  + components
  + config
  + features
     + a-feature
        - a-feature.android.js
        - a-feature.ios.js
        - a-feature.feature
     + definitions
        - default.common.js
        - default.android.js
        - default.ios.js
```

If I understood this correctly, inside the `features` folder, you can be flexible right now. The CLI has a scaffolding tool to create new features, but you can actually organize them in the way you wish :-) It is only required to use `*.android.js`, *.ios.js` and `*.common.js` as suffixes. 

In my team we created independent JavaScript files without be attached to features to be reused on other features :-) 

If I did not understood you well, could you give me an example of folder structure? :-)

Thanks for your comments, @mattapperson

**Comment by mattapperson** [comment]
@eridem ok, so for 2 and 3... I did not understand that this was how the cwd flag worked, but here is how I expressed things to work fwiw... not saying it's right:
- patata would consider the base directory to be wherever the patata command was run. 
- cwd flag could be used to alter the assumed base directory, overwriting the assumption that it should be the dir that the command was run in
- patata.yml is not required, but if it existed, patata would just look for it in whatever directory it was run.
- there would be a flag/config (including being able to set it in `patata.yml`)  for where to look for the components, config, features etc... This way `patata.yml` can exist in root but not everything has to

Your solution gets the job done, it is just not what I expected... 

as for 4... here is what I am shooting for:
```
+ root
  + integration-tests
    + components
    + config
    + features
        - a-feature.feature
        - b-feature.feature
        - c-feature.feature
     + definitions
        - default.common.js
        - default.android.js
        - default.ios.js
        - b-feature.common.js
```
The reason is we have non-programer project managers writing our acceptance tests here... we want to give them a folder so they can create these without causing any confusion. We then have default generic definitions that we use for all projects, and our engineers add support for special cases when needed

**Comment by mattapperson** [comment]
also having gotten this "working", it does not seem to actually see my tests. 
I have an app, in root I run `patata --cwd tests/intigration/ run ./testProj.app/`
in `tests/intigration/` I ran `patata feature "A wonderful feature"` and then in the .feature file I added my test. in the `a-wonderful-feature.common.js` file I have some definitions...

But when I run `patata --cwd tests/intigration/ run ./testProj.app/` from root I get:
```
[Patata] Version 6.0.1. More info: http://patata.io
[Patata] Executing tests...
[Patata] Verifying platform conditions...
[Patata] Running Patata...
0 scenarios
0 steps
0m00.000s
[Patata] Everything alright! Happy testing!
```

Any ideas?

**Comment by mattapperson** [comment]
Note:
if I console.log includedFiles in run-generic.js I get:
```
[ 'features/a-wonderful-feature/a-wonderful-feature.common.js',
  'hooks/hooks.common.js',
  'hooks/hooks.ios.js' ]
```

**Comment by mattapperson** [comment]
OK... so it seems that if you don't include a tag it runs nothing, and I would assume it would run all tags then...

**Comment by mattapperson** [comment]
hmmm ok now it "runs" with I added a tag, but fails because the app is not running... it never starts like a simulator or anything...

**Comment by mattapperson** [comment]
Sorry for so many posts here 😞 just reporting as I see things...

so run-generic.js has the platform var set to lets say 'ios', that var is used to select the included files... but it also seems to wind up (no idea the code path TBH) in capabilities.js... am I missing a flag?
I set the udid flag and it crashes because there is no capability for 'ios'... does that make sense?

**Comment by eridem** [comment]
Hi @mattapperson,
Sorry for late answers. I think we may be in different time schedules.

-I agree that sometimes you would like that POs/Managers write the acceptance criteria there (I promote this in my company too). You can decide how to distribute your files right now inside the `features` folder. As your example, you can separata the implementation from the specifications. You example is valid and it should work on `Patata`, with the exception of using `*.android.feature`, `*.ios.feature` and `*.common.feature`. If you want to have specifications for both platform, the POs/Managers can use `my-feature.common.feature` so it works in both platforms.

```
+ root
  + integration-tests
    + components
    + config
    + features
        - a-feature.common.feature
        - b-feature.common.feature
        - c-feature.common.feature
        + definitions
          - default.common.js
          - default.android.js
          - default.ios.js
          - b-feature.common.js
```
So, my answer is that you can distribute your files as you wish :-) It is only important to keep the suffixes in the files. In my team, we have the same app for both iOS and Android, but some features were a bit different between platforms so that's why we separated them, keeping `common` in case the feature works in the same way in both platforms.

- Yes, `patata` works with `tags` better. But if you do not use tags, it should run all of them.

- We did few tests with iOS, I guess we need to improve better this platform. Even the intention was working on both platform, we were using it only in Android. We should revise the `iOS` platform from scratch.

**Comment by eridem** [comment]
@mattapperson , would you mind if we close this issue and we create multiple ones for each question? I feel we talk about different things in one thread all time XD

**Comment by mattapperson** [comment]
Heh agreed
